### PR TITLE
[HLSL] Implement elementwise popcount

### LINKED ISSUE #99094: Implement the `countbits` HLSL Function
> Note The SPIRV Work is completed https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp#L481C3-L482C63
> We just need to add the test case.

- [ ] Implement `countbits` clang builtin,
- [ ] Link `countbits` clang builtin with `hlsl_intrinsics.h`
- [ ] Add sema checks for `countbits` to `CheckHLSLBuiltinFunctionCall` in `SemaChecking.cpp`
- [ ] Add codegen for `countbits` to `EmitHLSLBuiltinExpr` in `CGBuiltin.cpp`
- [ ] Add codegen tests to `clang/test/CodeGenHLSL/builtins/countbits.hlsl`
- [ ] Add sema tests to `clang/test/SemaHLSL/BuiltIns/countbits-errors.hlsl`
- [ ] Create the `DXILOpMapping` of `int_ctpop` to  `31` in `DXIL.td`
- [ ] Create the  `countbits.ll` and `countbits_errors.ll` tests in `llvm/test/CodeGen/DirectX/`
- [ ] Create SPIR-V backend test case in `llvm/test/CodeGen/SPIRV/hlsl-intrinsics/countbits.ll`

## DirectX

| DXIL Opcode | DXIL OpName | Shader Model | Shader Stages |
| ----------- | ----------- | ------------ | ------------- |
| 31 | Countbits | 6.0 | () |

## SPIR-V

# [OpBitCount](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpBitCount):

## Description:
  
Count the number of set bits in an object.  
  
Results are computed per component.  
  
*Result Type* must be a scalar or vector of [*integer type*](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Integer).
The components must be wide enough to hold the unsigned *Width* of
*Base* as an unsigned value. That is, no sign bit is needed or counted
when checking for a wide enough result width.  
  
*Base* must be a scalar or vector of [*integer type*](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Integer). It must
have the same number of components as *Result Type*.  
  
The result is the unsigned value that is the number of bits in *Base*
that are 1.

<table>
<colgroup>
<col style="width: 20%" />
<col style="width: 20%" />
<col style="width: 20%" />
<col style="width: 20%" />
<col style="width: 20%" />
</colgroup>
<thead>
<tr>
<th>Word Count</th>
<th>Opcode</th>
<th>Results</th>
<th>Operands</th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td class="tableblock halign-left valign-top"><p>4</p></td>
<td class="tableblock halign-left valign-top"><p>205</p></td>
<td
class="tableblock halign-left valign-top"><p><em>&lt;id&gt;</em><br />
<em>Result Type</em></p></td>
<td class="tableblock halign-left valign-top"><p><a
href="#ResultId"><em>Result &lt;id&gt;</em></a></p></td>
<td
class="tableblock halign-left valign-top"><p><em>&lt;id&gt;</em><br />
<em>Base</em></p></td>
</tr>
</tbody>
</table>

### Example 1
```hlsl
//dxc countbits_test.hlsl -T lib_6_8 -enable-16bit-types -O0

export uint4 fn(int4 p1) {
    return countbits(p1);
}
```
## HLSL:

Counts the number of bits (per component) set in the input integer.

## Syntax

``` syntax
uint countbits(
  in uint value
);
```

## Parameters

<dl> <dt>

*value* \[in\]
</dt> <dd>

Type: **uint**

The input value.

</dd> </dl>

## Return value

Type: **uint**

The number of bits.

## Remarks

The following overloaded versions are also available:

``` syntax
uint count_bits(uint value);
uint2 count_bits(uint2 value);
uint3 count_bits(uint3 value);
uint4 count_bits(uint4 value);
```

### Minimum Shader Model

This function is supported in the following shader models.

| Shader Model                                                                | Supported |
|-----------------------------------------------------------------------------|-----------|
| [Shader Model 5](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/d3d11-graphics-reference-sm5.md) and higher shader models | yes       |

 

This function is supported in the following types of shaders:

| Vertex | Hull | Domain | Geometry | Pixel | Compute |
|--------|------|--------|----------|-------|---------|
| x      | x    | x      | x        | x     | x       |

 

## See also

<dl> <dt>

[Intrinsic Functions](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-intrinsic-functions.md)
</dt> <dt>

[Shader Model 5](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/d3d11-graphics-reference-sm5.md)
</dt> </dl>

### PR DESCRIPTION
**Pull request by spall**
Add new elementwise popcount builtin to support HLSL function 'countbits'. 
elementwise popcount only accepts integer types.
Add hlsl intrinsic 'countbits'
Closes #99094

### CODE REVIEW

**Comment by spall** [comment]
This pull request is missing the appropriate text in 'ReleaseNotes.rst'. I am looking for advice on what is appropriate to put in that file, and where.

**Comment by python3kgae** [comment]
> This pull request is missing the appropriate text in 'ReleaseNotes.rst'. I am looking for advice on what is appropriate to put in that file, and where.

Here's a PR with ReleaseNotes.rst change https://github.com/llvm/llvm-project/commit/57f879cdd4c63189c569bdc1ce4e87e7342eea46#diff-ec770381d76c859f5f572db789175fe44410a72608f58ad5dbb14335ba56eb97

File: llvm/test/CodeGen/DirectX/countbits_error.ll
Comment (farzonl) [inline llvm/test/CodeGen/DirectX/countbits_error.ll:6]: should we test for `float`?

File: llvm/test/CodeGen/DirectX/countbits.ll
Comment (farzonl) [inline llvm/test/CodeGen/DirectX/countbits.ll:13]: It doesn't look like this test uses the `Function Attrs`.  If we don't then please remove them.

File: llvm/test/CodeGen/DirectX/countbits.ll
Comment (farzonl) [inline llvm/test/CodeGen/DirectX/countbits.ll:8]: NIT: There is usually a space after `CHECK:`
```suggestion
; CHECK: call i16 @dx.op.unary.i16(i32 31, i16 %{{.*}})
```

**Comment by spall** [comment]
> > This pull request is missing the appropriate text in 'ReleaseNotes.rst'. I am looking for advice on what is appropriate to put in that file, and where.
> 
> Here's a PR with ReleaseNotes.rst change [57f879c#diff-ec770381d76c859f5f572db789175fe44410a72608f58ad5dbb14335ba56eb97](https://github.com/llvm/llvm-project/commit/57f879cdd4c63189c569bdc1ce4e87e7342eea46#diff-ec770381d76c859f5f572db789175fe44410a72608f58ad5dbb14335ba56eb97)

Yeah I saw this example but am unsure where to put my text because the contents of that file are not the same now.

**Comment by python3kgae** [comment]
> > > This pull request is missing the appropriate text in 'ReleaseNotes.rst'. I am looking for advice on what is appropriate to put in that file, and where.
> > 
> > 
> > Here's a PR with ReleaseNotes.rst change [57f879c#diff-ec770381d76c859f5f572db789175fe44410a72608f58ad5dbb14335ba56eb97](https://github.com/llvm/llvm-project/commit/57f879cdd4c63189c569bdc1ce4e87e7342eea46#diff-ec770381d76c859f5f572db789175fe44410a72608f58ad5dbb14335ba56eb97)
> 
> Yeah I saw this example but am unsure where to put my text because the contents of that file are not the same now. Maybe the answer is to add my own section.

I see. Maybe we add HLSL Language Changes like C++ Language Changes at https://github.com/llvm/llvm-project/blob/main/clang/docs/ReleaseNotes.rst?plain=1#L106

**Comment by RKSimon** [comment]
I'd recommend you add it to the "C++ Language Changes" section.

**Review by RKSimon — COMMENTED** [review]
Please can you fix the clang-format warnings?

File: clang/lib/Sema/SemaHLSL.cpp
Comment (farzonl) [inline clang/lib/Sema/SemaHLSL.cpp:1525]: It doesn't seem like this function is used? If it isn't please remove.

**Comment by farzonl** [comment]
I don't see a `clang/test/CodeGenHLSL/builtins/countbits.hlsl`

**Comment by farzonl** [comment]
Also we probably still want a `clang/test/SemaHLSL/BuiltIns/countbits-errors.hlsl` to check the errors for float representations.

**Comment by farzonl** [comment]
Code looks fine, I need to see more tests before I can sign off.

File: clang/test/Sema/builtins-elementwise-math.c
Comment (RKSimon) [inline clang/test/Sema/builtins-elementwise-math.c:526]: please can you add tests for vector element count mismatches and implicit arg vs result vector extension/truncation/sign changes

File: clang/test/CodeGenHLSL/builtins/countbits.hlsl
Comment (farzonl) [inline clang/test/CodeGenHLSL/builtins/countbits.hlsl:6]: i dunno if the `// CHECK: define` add any value stopping at the `@`. I think before we were using them to check the function name, but that can be better done with a `CHECK-LABEL`.

File: llvm/test/CodeGen/DirectX/countbits.ll
Comment (farzonl) [inline llvm/test/CodeGen/DirectX/countbits.ll:1]: so I don't have to clean this up later can you change this to 
```suggestion
; RUN: opt -S -scalarizer -dxil-op-lower -mtriple=dxil-pc-shadermodel6.3-library %s | FileCheck %s
```
And then add a vector test.

File: llvm/test/CodeGen/DirectX/countbits.ll
Comment (farzonl) [inline llvm/test/CodeGen/DirectX/countbits.ll:25]:
```suggestion
define noundef <4 x i32> @countbits_vec4_i32(<4 x i32> noundef %a) #0 {
entry:
  ; CHECK: [[ee0:%.*]] = extractelement <4 x i32> %a, i64 0
  ; CHECK: [[ie0:%.*]] = call i32 @dx.op.unary.i32(i32 31, i32 [[ee0]])
  ; CHECK: [[ee1:%.*]] = extractelement <4 x i32> %a, i64 1
  ; CHECK: [[ie1:%.*]] = call i32 @dx.op.unary.i32(i32 31, i32 [[ee1]])
  ; CHECK: [[ee2:%.*]] = extractelement <4 x i32> %a, i64 2
  ; CHECK: [[ie2:%.*]] = call i32 @dx.op.unary.i32(i32 31, i32 [[ee2]])
  ; CHECK: [[ee3:%.*]] = extractelement <4 x i32> %a, i64 3
  ; CHECK: [[ie3:%.*]] = call i32 @dx.op.unary.i32(i32 31, i32 [[ee3]])
  ; CHECK: insertelement <4 x i32> poison, i32 [[ie0]], i64 0
  ; CHECK: insertelement <4 x i32> %{{.*}}, i32 [[ie1]], i64 1
  ; CHECK: insertelement <4 x i32> %{{.*}}, i32 [[ie2]], i64 2
  ; CHECK: insertelement <4 x i32> %{{.*}}, i32 [[ie3]], i64 3
  %2 = call <4 x i32> @llvm.ctpop.v4i32(<4 x i32> %a) 
  ret <4 x i32> %2
}
```

File: llvm/test/CodeGen/SPIRV/hlsl-intrinsics/countbits.ll
Comment (farzonl) [inline llvm/test/CodeGen/SPIRV/hlsl-intrinsics/countbits.ll:21]: add a vector test

File: llvm/test/CodeGen/DirectX/countbits.ll
Comment (farzonl) [inline llvm/test/CodeGen/DirectX/countbits.ll:28]:
```suggestion
declare i64 @llvm.ctpop.i64(i64)
declare <4 x i32> @llvm.ctpop.v4i32(<4 x i32>)
```

File: clang/test/SemaHLSL/BuiltIns/countbits-errors.hlsl
Comment (farzonl) [inline clang/test/SemaHLSL/BuiltIns/countbits-errors.hlsl:9]: I don't know if these notes add any value. There are cases where `-verify-ignore-unexpected` is warranted. This might be one of those cases.

**Review by farzonl — APPROVED** [review]

**Review by RKSimon — APPROVED** [review]
LGTM after @farzonl minors have been addressed - once this is committed I'm intending to use `__builtin_elementwise_popcount` to simplify a number of C++ target intrinsics that do the same thing.

**Pull request merged**
